pysd/py_backend/output.py

```diff
@@ -0,0 +1,426 @@
+            Whether to add a time as the first dimension for the variables.
+
+        Returns
+        -------
+        None
+
+        """
+
+        for key in capture_elements:
+            comp = getattr(model.components, key)
+            comp_vals = comp()
+
+            dims = ()
+
+            if isinstance(comp_vals, (xr.DataArray, np.ndarray)):
+                if comp.subscripts:
+                    dims = tuple(comp.subscripts)
+
+            if add_time:
+                dims = ("time",) + dims
+
+            self.ds.createVariable(key, "f8", dims, compression="zlib")
+
+            # adding units and description as metadata for each var
+            self.ds[key].units = model.doc.loc[
+                        model.doc["Py Name"] == key,
+                        "Units"].values[0] or "Missing"
+            self.ds[key].description = model.doc.loc[
+                        model.doc["Py Name"] == key,
+                        "Comment"].values[0] or "Missing"
+
+
+class DataFrameHandler(OutputHandlerInterface):
+    """
+    Manages simulation results stored as pandas DataFrame.
+    """
+    def __init__(self, out_file):
+        self.ds = None
+        self.output_file = out_file
+
+    def initialize(self, model, capture_elements):
+        """
+        Creates a pandas DataFrame and adds model variables as columns.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+
+        """
+        self.ds =  pd.DataFrame(columns=capture_elements)
+
+    def update(self, model, capture_elements):
+        """
+        Add a row to the results pandas DataFrame with the values of the
+        variables listed in capture_elements.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+        """
+
+        self.ds.at[model.time.round()] = [
+            getattr(model.components, key)()
+            for key in capture_elements]
+
+    def postprocess(self, **kwargs):
+        """
+        Delete time column from the pandas DataFrame and flatten xarrays if
+        required.
+
+        Returns
+        -------
+        ds: pandas.DataFrame
+            Simulation results stored as a pandas DataFrame.
+        """
+        # delete time column as it was created only for avoiding errors
+        # of appending data. See previous TODO.
+        del self.ds["time"]
+
+        # enforce flattening if df is to be saved to csv or tab file
+        flatten = True if self.output_file else kwargs.get("flatten", None)
+
+        df = utils.make_flat_df(self.ds,
+                                kwargs["return_addresses"],
+                                flatten)
```

Comment:
Maybe we can migrate the function `make_flate_df` to be a private method of this class, which will be much cleaner :)
I think it is not used in more places than by the output df from the integration. We may have to update the tests.

Reply:
Done!